### PR TITLE
Add MIB downloader to generator

### PR DESCRIPTION
**Pull request by SuperQ**
Add a Makefile to automatically download all the required MIBs.

Signed-off-by: Ben Kochie <superq@gmail.com>

### CODE REVIEW

File: generator/Makefile [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+PALOALTO_URL   := 'https://www.paloaltonetworks.com/content/dam/pan/en_US/assets/zip/technical-documentation/snmp-mib-modules/PAN-MIB-MODULES-7.0.zip'
+
+
+.DEFAULT: all
+
+.PHONY: all
+all: mibs
+
+clean:
+	@rm -rv $(MIBDIR)
```
Comment (brian-brazil) [inline generator/Makefile]: I don't think it's wise to blow away the user's mibdir like this, it could contain other mibs.

Comment (SuperQ) [reply generator/Makefile]: Sounds reasonable.

**Comment by SuperQ** [comment]
FYI, I left out UCD-SNMP-MIB, it's included in `/usr/share/snmp/mibs`.

**Comment by SuperQ** [comment]
I've added a generator build and snmp.yml diff test to CircleCI.

File: generator/.gitignore [hunk 2]
```diff
@@ -0,0 +1 @@
+mibs
```
Comment (brian-brazil) [inline generator/.gitignore]: Can we put this in the main gitignore?

File: generator/Makefile [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	@curl -s -o $(MIBDIR)/ARISTA-SW-IP-FORWARDING-MIB -L "$(ARISTA_URL)/ARISTA-SW-IP-FORWARDING-MIB.txt"
+
+$(MIBDIR)/.cisco_v2:
+	@echo ">> Downloading cisco v2"
+	@curl -s -L $(CISCO_URL) \
+		| tar -C $(MIBDIR) --strip-components=3 -zxv
+	@touch $(MIBDIR)/.cisco_v2
+
+$(MIBDIR)/mikrotik.mib:
+	@echo ">> Downloading cisco v2"
```
Comment (brian-brazil) [inline generator/Makefile]: Message is wrong

File: .circleci/config.yml [hunk 4]
```diff
@@ -28,6 +28,25 @@ jobs:
...
+    resource_class: large
+
+    environment:
+      MIBDIRS: mibs:/usr/share/snmp/mibs
+    steps:
+    - checkout:
+        path: /go/src/github.com/prometheus/snmp_exporter
+    - run: sudo sh -c "echo 'deb http://deb.debian.org/debian stable non-free' > /etc/apt/sources.list.d/non-free.list"
+    - run: sudo apt-get update
+    - run: sudo apt-get -y install build-essential diffutils libsnmp-dev snmp-mibs-downloader
```
Comment (brian-brazil) [inline .circleci/config.yml]: What do we need snmp-mibs-downloader for? I'd expect the MIBs to be self-contained.

File: generator/Makefile [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+$(MIBDIR)/ARISTA-SMI-MIB:
+	@echo ">> Downloading ARISTA-SMI-MIB"
+	@curl -s -o $(MIBDIR)/ARISTA-SMI-MIB -L "$(ARISTA_URL)/ARISTA-SMI-MIB.txt"
+
+$(MIBDIR)/ARISTA-SW-IP-FORWARDING-MIB:
+	@echo ">> Downloading ARISTA-SW-IP-FORWARDING-MIB"
+	@curl -s -o $(MIBDIR)/ARISTA-SW-IP-FORWARDING-MIB -L "$(ARISTA_URL)/ARISTA-SW-IP-FORWARDING-MIB.txt"
+
+$(MIBDIR)/.cisco_v2:
+	@echo ">> Downloading cisco v2"
```
Comment (RichiH) [inline generator/Makefile]: `cisco_v2`

File: generator/Makefile [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+  $(MIBDIR)/UBNT-MIB \
+  $(MIBDIR)/UBNT-UniFi-MIB \
+  $(MIBDIR)/UBNT-AirMAX-MIB.txt
+
+mib-dir:
+	@mkdir -p -v $(MIBDIR)
+
+$(MIBDIR)/apc-powernet-mib:
+	@echo ">> Downloading apc-powernet-mib"
+	@curl -s -o $(MIBDIR)/apc-powernet-mib -L $(APC_URL)
```
Comment (RichiH) [inline generator/Makefile]: No check if curl is installed.

File: generator/Makefile [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+$(MIBDIR)/servertech-sentry3-mib:
+	@echo ">> Downloading servertech-sentry3-mib"
+	@curl -s -o $(MIBDIR)/servertech-sentry3-mib -L $(SERVERTECH_URL)
+
+$(MIBDIR)/.synology:
+	$(eval TMP := $(shell mktemp))
+	@echo ">> Downloading synology to $(TMP)"
+	@curl -s -o $(TMP) -L $(SYNOLOGY_URL)
+	@unzip -j -d $(MIBDIR) $(TMP)
```
Comment (RichiH) [inline generator/Makefile]: No check if unzip is installed.

File: generator/Makefile [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+	@rm -v $(TMP)
+	@touch $(MIBDIR)/.paloalto_panos
+
+$(MIBDIR)/servertech-sentry3-mib:
+	@echo ">> Downloading servertech-sentry3-mib"
+	@curl -s -o $(MIBDIR)/servertech-sentry3-mib -L $(SERVERTECH_URL)
+
+$(MIBDIR)/.synology:
+	$(eval TMP := $(shell mktemp))
```
Comment (RichiH) [inline generator/Makefile]: If that fails, `$TMP` is just empty.

**Comment by RichiH** [comment]
> FYI, I left out UCD-SNMP-MIB, it's included in `/usr/share/snmp/mibs`.

Which mibs dir? I doubt there is a unix-wide standard to rely on.

File: generator/Makefile (hunk 8 above)
Comment (SuperQ) [reply generator/Makefile]: Make should fail if mktemp returns a non-zero code.

File: generator/Makefile (hunk 7 above)
Comment (SuperQ) [reply generator/Makefile]: I consider this out of scope, Make will handle the command failure.

File: generator/Makefile (hunk 6 above)
Comment (SuperQ) [reply generator/Makefile]: Also out of scope, Make handles this.

**Comment by SuperQ** [comment]
@RichiH It turns out we don't need `UCD-SNMP-MIB` at all for the current generator.yml.

**Comment by RichiH** [comment]
Generally speaking, I would like to try and avoid depending on any mib which has not been downloaded.

**Comment by SuperQ** [comment]
@RichiH Yes, previously we used the `snmp-mibs-downloader` package, but it turns out to be unnecessary. As we add new features to the generator.yml, we can add more MIBs to the downloader.

**Comment by RichiH** [comment]
I think we can merge current state.

General comment: MIB names are repeated so often, I wonder if it wouldn't make sense to loop through them.

**Pull request merged**